Rework timer message to be driven by catchup

Geod24 opened the pull request:
That should make running a full node much better. We still get a flood of message at startup, but not during continous operation.

However, it also shows one issue with our timers:
```
2022-02-17 15:10:21,463 Info [agora.consensus.state.Ledger] - Completed externalization of block #1650
2022-02-17 15:10:58,37 Info [agora.node.FullNode] - Ledger has fell behind, missing 465 blocks (current height: 1650)
2022-02-17 15:12:05,340 Info [agora.node.FullNode] - Ledger has fell behind, missing 467 blocks (current height: 1650)
2022-02-17 15:13:12,264 Info [agora.node.FullNode] - Ledger has fell behind, missing 468 blocks (current height: 1650)
2022-02-17 15:14:21,539 Info [agora.node.FullNode] - Ledger has fell behind, missing 469 blocks (current height: 1650)
```

Look at the timing. I'm using the default config, so it should trigger every 20 seconds.

Geod24 committed:
Trivial: Add missing attributes to Ledger.expectedHeight

Geod24 committed:
Rework timer message to be driven by catchup

The previous implementation was frequently showing the network discovery message,
which didn't make much sense to users and developers alike.
This downgrade that message to Trace level, and instead try to intelligently
print the current state of the Ledger compared to the expected state.

hewison-chris commented on source/agora/node/FullNode.d:
```diff
@@ -545,11 +545,14 @@ public class FullNode : API
-        if (this.network.peers.empty())  // no clients yet (discovery)
+        if (this.network.peers.empty())
+        {
+            // TODO: Iff the node is the single validator for the network,
```
```suggestion
            // TODO: If the node is the single validator for the network,
```

hewison-chris replied:
If it is the only validator then the message is still correct.

Geod24 replied:
`Iff` is correct isn't it ?
Also the message is correct, but it should not be a warning. It shouldn't even be logged IMO.

hewison-chris replied:
ok so `Iff` as in `If and only if`

hewison-chris commented on source/agora/node/FullNode.d:
```diff
@@ -558,9 +561,22 @@ public class FullNode : API
+                          this.ledger.height, expected);
+        else if (expected > this.ledger.height)
+        {
+            this.log.info("Ledger has fell behind, missing {} blocks (current height: {})",
```
```suggestion
            this.log.info("Ledger has fallen behind, missing {} blocks (current height: {})",
```

Geod24 replied:
[I should have known](https://www.youtube.com/watch?v=9BXNKyoW_Ow).

hewison-chris reviewed (changes requested):
LGTM

Geod24 commented:
Updated.
```
2022-02-17 15:34:33,674 Info [agora.node.FullNode] - Ledger out of sync, missing 489 blocks (current height: 1650, delay: 8 hours and 9 minutes)
2022-02-17 15:35:42,59 Info [agora.node.FullNode] - Ledger out of sync, missing 490 blocks (current height: 1650, delay: 8 hours and 10 minutes)
```

linked0 reviewed (approved):
LGTM.
By the way, is this applied to the code?
```
2022-02-17 15:35:42,59 Info [agora.node.FullNode] - Ledger out of sync, missing 490 blocks (current height: 1650, delay: 8 hours and 10 minutes)
```

hewison-chris reviewed (approved):
LGTM

Geod24 commented:
@linked0 : Yes, that was me running a TestNet full node.

Geod24 committed:
Trivial: Remove assert message redundant with checkaction=context

Geod24 committed:
Test: Wait longer for assertSameBlocks

The nodes might take a bit longer than 5 seconds to catch up.
Once they 'getBlockHeight' is correct, the getBlocksFrom should not fail,
so we can leave it at 5 seconds (or even remove the retry).

Geod24 commented:
> node-0_1  | 2022-02-20 21:41:21,787 Info [agora.node.FullNode] - Ledger out of sync, missing 7186807 blocks (current height: 9, delay: 59 weeks, 2 days, 21 hours, 40 minutes, and 35 secs)

Maybe we should fix the integration test...
